src/main/scala/org/apache/spark/shuffle/ucx/NvkvShuffleMapOutputWriter.scala

```diff
@@ -0,0 +1,291 @@
+    if (reducePartitionId <= lastPartitionId) throw new IllegalArgumentException("Partitions should be requested in increasing order.")
+    lastPartitionId = reducePartitionId
+    currChannelPosition = getBlockOffset + bytesWrittenToMergedFile + totalPartitionsPadding
+    /* 
+     * Temporary solution for distributing the load between the NVME devices to get full read BW.
+     * This solution is not optimal because is nt taking into account NUMA topology and it's dependent in randmoness.
+     * See SparkDPU documentation for more information. 
+     */
+    dsIdx = Random.nextInt(nvkvWrapper.getNumOfDevices)
+    NvkvShuffleMapOutputWriter.log.debug(s"NvkvShuffleMapOutputWriter reducePartition offset $currChannelPosition")
+    new NvkvShufflePartitionWriter(reducePartitionId)
+  }
+
+  @throws[IOException]
+  override def commitAllPartitions: Array[Long] = {
+    NvkvShuffleMapOutputWriter.log.debug("NvkvShuffleMapOutputWriter commitAllPartitions")
+    // Check the position after transferTo loop to see if it is in the right position and raise a
+    // exception if it is incorrect. The position will not be increased to the expected length
+    // after calling transferTo in kernel version 2.6.32. This issue is described at
+    // https://bugs.openjdk.java.net/browse/JDK-7052359 and SPARK-3948.
+
+    var blockOffset = getBlockOffset;
+
+    //TODO - move to shuffleTransport    
+    var packMapperData: ByteBuffer = ByteBuffer.allocateDirect(UnsafeUtils.INT_SIZE + // numOfMappers
+                                                               UnsafeUtils.INT_SIZE + // numOfReducePartitions
+                                                               UnsafeUtils.INT_SIZE + // mapperId
+                                                               2*UnsafeUtils.LONG_SIZE*partitionLengths.size // offset + length for each reducePartition
+                                                              ).order(ByteOrder.LITTLE_ENDIAN)
+    packMapperData.putInt(1)
+    packMapperData.putInt(partitionLengths.size)
+    packMapperData.putInt(mapId.toInt)
+
+    partitionLengths.zip(0 until partitionLengths.size).foreach{ 
+        case (partitionLength, reduceId) => {
+            NvkvShuffleMapOutputWriter.log.debug(s"shuffleId $shuffleId mapId $mapId reducerId $reduceId offset $blockOffset size $partitionLength")
+            NvkvShuffleMapOutputWriter.log.debug(s"Reduce partition stats: offset ${nvkvWrapper.getPartitonOffset(shuffleId, mapId, reduceId)} length ${nvkvWrapper.getPartitonLength(shuffleId, mapId, reduceId)} padding ${partitionsPadding(reduceId)}")
+            packMapperData.putLong(nvkvWrapper.getPartitonOffset(shuffleId, mapId, reduceId))
+            packMapperData.putLong(nvkvWrapper.getPartitonLength(shuffleId, mapId, reduceId))
+            blockOffset += partitionLength
+        }
+    }
+
+    packMapperData.rewind
```

Comment:
Should be changed to `packMapperData.rewind()` as no parenthesis suggests the function is just returning a value (without a side effect) which is not the case here